src/sql/src/query_model/hir.rs

```diff
@@ -0,0 +1,133 @@
+        if !self.model.get_box(box_id).borrow().is_select() {
+            box_id = self.wrap_within_select(box_id);
+        }
+        box_id
+    }
+
+    fn generate_internal(&mut self, expr: &HirRelationExpr) -> BoxId {
+        match expr {
+            // HirRelationExpr::Get { id, typ } => {
+            //     self.model.make_box(BoxType::BaseTable(BaseTable {}))
+            // }
+            HirRelationExpr::Constant { rows, typ } => {
+                assert!(typ.arity() == 0, "expressions are not yet supported",);
+                self.model.make_box(BoxType::Values(Values {
+                    rows: rows.iter().map(|_| Vec::new()).collect_vec(),
+                }))
+            }
+            HirRelationExpr::Map { input, scalars } => {
+                let box_id = self.generate_select(input);
+                // @todo self-referencing Maps
+                for scalar in scalars.iter() {
+                    let expr = self.generate_expr(scalar, box_id);
+                    let b = self.model.get_box(box_id);
+                    b.borrow_mut().columns.push(Column { expr, alias: None });
+                }
+                box_id
+            }
+            HirRelationExpr::Project { input, outputs } => {
+                let input_box_id = self.generate_internal(input);
+                let select_id = self.model.make_select_box();
+                let quantifier_id =
+                    self.model
+                        .make_quantifier(QuantifierType::Foreach, input_box_id, select_id);
+                let mut select_box = self.model.get_box(select_id).borrow_mut();
+                for position in outputs {
+                    select_box.columns.push(Column {
+                        expr: BoxScalarExpr::ColumnReference(ColumnReference {
+                            quantifier_id,
+                            position: *position,
+                        }),
+                        alias: None,
+                    });
+                }
+                select_id
+            }
+            _ => panic!("unsupported expression type"),
+        }
+    }
+
+    /// Returns a Select box ranging over the given box, projecting
+    /// all of its columns.
+    fn wrap_within_select(&mut self, box_id: BoxId) -> BoxId {
+        let select_id = self.model.make_select_box();
+        self.model
+            .make_quantifier(QuantifierType::Foreach, box_id, select_id);
+        let mut select_box = self.model.get_box(select_id).borrow_mut();
+        select_box.add_all_input_columns(&self.model);
+        select_id
+    }
+
+    fn generate_expr(&mut self, expr: &HirScalarExpr, context_box: BoxId) -> BoxScalarExpr {
+        match expr {
+            HirScalarExpr::Literal(row, col_type) => {
+                BoxScalarExpr::Literal(row.clone(), col_type.clone())
+            }
+            HirScalarExpr::Column(c) => {
+                assert!(c.level == 0, "correlated columns not yet supported");
```

Comment:
change this to assert_eq!(c.level, 0, "correlated columns ...") so that the level prints out if the equality fails.

Reply:
This `assert` is removed by the following patch in https://github.com/MaterializeInc/materialize/pull/8969, so it won't last very long